include/boost/hana/difference.hpp

```diff
@@ -21,11 +22,21 @@ BOOST_HANA_NAMESPACE_BEGIN
     //! @cond
     template <typename Xs, typename Ys>
     constexpr auto difference_t::operator()(Xs&& xs, Ys&& ys) const {
-        using S = typename hana::tag_of<Xs>::type;
-        using Difference = BOOST_HANA_DISPATCH_IF(difference_impl<S>,
-            true
+        using TX = typename hana::tag_of<Xs>::type;
+        using TY = typename hana::tag_of<Ys>::type;
+        using Difference = BOOST_HANA_DISPATCH_IF(difference_impl<TX>,
+            hana::SetTheoretic<TX>::value &&
+            std::is_same<TX, TY>::value
         );
 
+    #ifndef BOOST_HANA_CONFIG_DISABLE_CONCEPT_CHECKS
+        static_assert(hana::SetTheoretic<TX>::value,
+        "hana::difference(xs, ys) requires 'xs' to be SetTheoretic");
+
+        static_assert(std::is_same<TX, TY>::value,
+        "hana::difference(xs, ys) requires 'xs' and 'ys' to be SetTheoretic");
```

Comment:
The static_assert message is wrong; it should say "requires 'xs' and 'ys' to be of the same data type", or equivalent.

Reply:
fixed